File: src/bgremover/stages/trimap.py
"""Stage 3 — uncertainty band derivation (pseudo-trimap).

Classical matting demanded a hand-painted trimap. We derive one.

Two sources of "unknown", unioned:

  a) Low model confidence. Wherever the segmentation probability is neither
     near 0 nor near 1, the model is telling us it doesn't know. This is the
     better signal and it is what BEN2 calls Confidence Guided Matting.
  b) A guaranteed minimum band around the mask boundary. A very confident
     model returns a razor-thin ambiguous region, but matting needs room to
     solve in. Without this, hair that the model called "definitely
     background" never gets a chance to come back.

Only the unknown band goes to stage 4, which is the expensive one, so band
width is the main quality/speed dial in the whole pipeline.
"""

from __future__ import annotations

import numpy as np
from scipy import ndimage

from ..types import TRIMAP_BG, TRIMAP_FG, TRIMAP_UNKNOWN

REFERENCE_EDGE = 1024
"""Short edge, in px, that `band_width` is quoted against. See `derive`."""


def derive(
    prob: np.ndarray,
    *,
    fg_threshold: float = 0.95,
    bg_threshold: float = 0.05,
    band_width: int = 12,
    absolute_band: bool = False,
) -> np.ndarray:
    """Build a trimap from a segmentation probability map.

    Args:
        prob: (H, W) float in [0, 1] — stage 2 output at full working res.
        fg_threshold: at or above this, treat as definite foreground.
        bg_threshold: at or below this, treat as definite background.
        band_width: minimum unknown band either side of the boundary, in px
            **at a REFERENCE_EDGE short edge**, rescaled to the actual image.
            0 disables the morphological band and trusts confidence alone.
        absolute_band: take `band_width` as literal pixels, skipping the
            rescale. For callers that already know their working resolution.

    A fixed pixel band is wrong, and wrong in both directions: 12 px is a
    hairline on a 4000 px product shot and 12% of the frame on a 208 px
    thumbnail. Measured on a 208x242 image, an unscaled 12 px band put 41% of
    all pixels in the unknown region — stage 4 was being asked to re-segment
    the picture, and it answered by softening every edge into a halo. Scaled,
    the same call yields 8%.

    Returns:
        (H, W) uint8 trimap using TRIMAP_BG / TRIMAP_UNKNOWN / TRIMAP_FG.
    """
    if prob.ndim != 2:
        raise ValueError(f"prob must be 2-D (H, W), got shape {prob.shape}")
    if not 0.0 <= bg_threshold < fg_threshold <= 1.0:
        raise ValueError(
            f"need 0 <= bg_threshold ({bg_threshold}) < fg_threshold ({fg_threshold}) <= 1"
        )
    if band_width < 0:
        raise ValueError(f"band_width must be >= 0, got {band_width}")

    prob = np.clip(prob.astype(np.float32), 0.0, 1.0)

    definite_fg = prob >= fg_threshold
    definite_bg = prob <= bg_threshold

    # (a) confidence-driven unknown
    unknown = ~(definite_fg | definite_bg)

    # (b) morphological band around the boundary of the thresholded mask
    iterations = _scaled_band(band_width, prob.shape, absolute_band)
    if iterations > 0:
        solid = prob >= 0.5
        outer = ndimage.binary_dilation(solid, iterations=iterations)
        inner = ndimage.binary_erosion(solid, iterations=iterations)
        unknown |= outer & ~inner

    trimap = np.full(prob.shape, TRIMAP_BG, dtype=np.uint8)
    trimap[definite_fg] = TRIMAP_FG
    trimap[unknown] = TRIMAP_UNKNOWN
    return trimap


def _scaled_band(band_width: int, shape: tuple[int, ...], absolute: bool) -> int:
    """Convert a REFERENCE_EDGE-relative band width to pixels for this image."""
    if band_width <= 0:
        return 0
    if absolute:
        return band_width
    short_edge = min(shape[:2])
    return max(1, round(band_width * short_edge / REFERENCE_EDGE))


def band_fraction(trimap: np.ndarray) -> float:
    """Share of pixels in the unknown band. The stage-4 cost driver.

    Useful as a guard: a band over ~0.35 usually means stage 2 failed and
    matting is about to be asked to segment the image from scratch, which it
    cannot do.
    """
    return float(np.count_nonzero(trimap == TRIMAP_UNKNOWN) / trimap.size)


def to_pymatting(trimap: np.ndarray) -> np.ndarray:
    """Convert to pymatting's convention: float64, 0 = bg, 1 = fg, else unknown."""
    out = np.zeros(trimap.shape, dtype=np.float64)
    out[trimap == TRIMAP_FG] = 1.0
    out[trimap == TRIMAP_UNKNOWN] = 0.5
    return out

File: web/src/stages/trimap.ts
/**
 * Stage 3 — uncertainty band derivation.
 *
 * Port of the Python stage. Two sources of "unknown", unioned:
 *   a) low model confidence, which is the better signal;
 *   b) a guaranteed minimum band around the boundary, because a confident
 *      model returns a razor-thin ambiguous region and matting needs room.
 *
 * Only the band goes to stage 4, so band width is the main quality/speed
 * dial. It is quoted in px at REFERENCE_EDGE and rescaled — an absolute
 * pixel band is a hairline on a 4000px photo and 12% of a thumbnail.
 */

import { TRIMAP_BG, TRIMAP_FG, TRIMAP_UNKNOWN, type Plane } from "../types";

export const REFERENCE_EDGE = 1024;

export interface TrimapOptions {
  fgThreshold?: number;
  bgThreshold?: number;
  /** px at REFERENCE_EDGE; rescaled to the actual image. */
  bandWidth?: number;
}

export function derive(prob: Plane, opts: TrimapOptions = {}): Uint8Array {
  const { fgThreshold = 0.95, bgThreshold = 0.05, bandWidth = 12 } = opts;
  if (!(bgThreshold >= 0 && bgThreshold < fgThreshold && fgThreshold <= 1)) {
    throw new Error(
      `need 0 <= bgThreshold (${bgThreshold}) < fgThreshold (${fgThreshold}) <= 1`,
    );
  }

  const { data, width: w, height: h } = prob;
  const n = w * h;
  const out = new Uint8Array(n);

  const iterations = scaledBand(bandWidth, w, h);

  // Pass 1: confidence.
  const solid = new Uint8Array(n);
  for (let i = 0; i < n; i++) {
    const p = data[i];
    solid[i] = p >= 0.5 ? 1 : 0;
    out[i] = p >= fgThreshold ? TRIMAP_FG : p <= bgThreshold ? TRIMAP_BG : TRIMAP_UNKNOWN;
  }

  // Pass 2: morphological band, as dilate(solid) AND NOT erode(solid).
  // Both are separable with a square structuring element, so this is two
  // 1-D passes each rather than a full 2-D neighbourhood scan.
  if (iterations > 0) {
    const outer = morph(solid, w, h, iterations, true);
    const inner = morph(solid, w, h, iterations, false);
    for (let i = 0; i < n; i++) {
      if (outer[i] === 1 && inner[i] === 0) out[i] = TRIMAP_UNKNOWN;
    }
    // Confidence must not be overridden by the band for definite pixels that
    // the band happens to miss; re-assert FG where the model was certain.
    for (let i = 0; i < n; i++) {
      if (data[i] >= fgThreshold && out[i] !== TRIMAP_UNKNOWN) out[i] = TRIMAP_FG;
    }
  }
  return out;
}

export function scaledBand(bandWidth: number, w: number, h: number): number {
  if (bandWidth <= 0) return 0;
  return Math.max(1, Math.round((bandWidth * Math.min(w, h)) / REFERENCE_EDGE));
}

export function bandFraction(trimap: Uint8Array): number {
  let n = 0;
  for (let i = 0; i < trimap.length; i++) if (trimap[i] === TRIMAP_UNKNOWN) n++;
  return n / trimap.length;
}

/** Separable square dilation (`dilate`) or erosion, radius `r`. */
function morph(src: Uint8Array, w: number, h: number, r: number, dilate: boolean): Uint8Array {
  const pick = dilate ? Math.max : Math.min;
  const tmp = new Uint8Array(w * h);
  const dst = new Uint8Array(w * h);
  const edge = dilate ? 0 : 1; // outside the image: empty for dilate, solid for erode

  for (let y = 0; y < h; y++) {
    const row = y * w;
    for (let x = 0; x < w; x++) {
      let v = edge;
      for (let k = -r; k <= r; k++) {
        const xx = x + k;
        const s = xx < 0 || xx >= w ? edge : src[row + xx];
        v = pick(v, s);
      }
      tmp[row + x] = v;
    }
  }
  for (let x = 0; x < w; x++) {
    for (let y = 0; y < h; y++) {
      let v = edge;
      for (let k = -r; k <= r; k++) {
        const yy = y + k;
        const s = yy < 0 || yy >= h ? edge : tmp[yy * w + x];
        v = pick(v, s);
      }
      dst[y * w + x] = v;
    }
  }
  return dst;
}
